### ISSUE TITLE
Design for Rook software upgrades

### ISSUE THREAD

**Issue by hunter** [rook/rook#277]
I thought I'd test out upgrading from 0.1.1 to 0.2.0 and ran into a few issues. Since Daemonsets don't handle rolling updates, I deleted the Daemonset (which also deleted the pods). After launching with the 0.2.0 manifest, I get the following error:

```
kubectl logs po/rookd-8fzjs -n ceph
2016-12-09 01:47:28.725481 I | util: loaded id from data dir
2016-12-09 01:47:28.725562 I | clusterd: Starting cluster. configDir=/var/lib/rook, nodeID=7c07d74f3f2e, url=https://discovery.etcd.io/e6240dc83f2c8bbebbde477e991180e8, members=, publicIPv4=10.2.45.6, privateIPv4=10.2.45.6, logLevel=INFO
2016-12-09 01:47:29.866268 I | etcdboot: cluster max size is: 1
2016-12-09 01:47:30.144000 I | etcdboot: currentNodes: [http://10.2.31.13:53379]
2016-12-09 01:47:30.144028 I | etcdboot: current etcd cluster nodes: [http://10.2.31.13:53379]
2016-12-09 01:47:30.144035 I | etcdboot: current localURL: http://10.2.45.6:53379
2016-12-09 01:47:30.144042 I | etcdboot: quorum is already formed, returning current cluster members: [http://10.2.31.13:53379]
dial tcp 10.2.31.13:53379: getsockopt: no route to host
```

I assume this is related to the use of an emptydir for `/var/lib/rook` which gets cleared when deleting the pods. The manifest is pretty basic and wouldn't be ideal in a production environment, so would be good to work through making it more robust and handle the upgrades, scaling, etc.

**Comment by hunter** [comment]
Just to add to that. Although not unexpected since the devices were already formatted, after generating a new etcd token, the following error is returned:

```
$ kubectl logs po/rookd-xb7xs
2016-12-09 08:41:11.565796 I | util: loaded id from data dir
2016-12-09 08:41:11.567203 I | clusterd: Starting cluster. configDir=/var/lib/rook, nodeID=e57155ccfa35, url=https://discovery.etcd.io/f8a6ff340e317feb8b2551dcdc10d370, members=, publicIPv4=10.2.45.7, privateIPv4=10.2.45.7, logLevel=INFO
2016-12-09 08:41:12.744369 I | etcdboot: cluster max size is: 1
2016-12-09 08:41:13.013718 I | etcdboot: currentNodes: [http://10.2.8.26:53379 http://10.2.31.16:53379 http://10.2.13.29:53379]
2016-12-09 08:41:13.013752 I | etcdboot: current etcd cluster nodes: [http://10.2.8.26:53379 http://10.2.31.16:53379 http://10.2.13.29:53379]
2016-12-09 08:41:13.013759 I | etcdboot: current localURL: http://10.2.45.7:53379
2016-12-09 08:41:13.013766 I | etcdboot: quorum is already formed, returning current cluster members: [http://10.2.8.26:53379 http://10.2.31.16:53379 http://10.2.13.29:53379]
2016-12-09 08:41:13.021948 I | exec: Running command: lsblk --all -n -l --output KNAME
2016-12-09 08:41:13.026144 I | exec: Running command: lsblk /dev/sda -b -d -P -o SIZE,ROTA,RO,TYPE,PKNAME
2016-12-09 08:41:13.029899 I | exec: Running command: sgdisk -p /dev/sda
2016-12-09 08:41:13.041258 I | exec: Running command: df --output=source,fstype
2016-12-09 08:41:13.051514 I | exec: Running command: lsblk /dev/sda1 -b -d -P -o SIZE,ROTA,RO,TYPE,PKNAME
2016-12-09 08:41:13.053657 I | exec: Running command: df --output=source,fstype
2016-12-09 08:41:13.061730 I | exec: Running command: lsblk /dev/sda2 -b -d -P -o SIZE,ROTA,RO,TYPE,PKNAME
2016-12-09 08:41:13.063605 I | exec: Running command: df --output=source,fstype
2016-12-09 08:41:13.073353 I | exec: Running command: lsblk /dev/sda3 -b -d -P -o SIZE,ROTA,RO,TYPE,PKNAME
2016-12-09 08:41:13.075201 I | exec: Running command: df --output=source,fstype
2016-12-09 08:41:13.082708 I | exec: Running command: lsblk /dev/sda4 -b -d -P -o SIZE,ROTA,RO,TYPE,PKNAME
2016-12-09 08:41:13.084592 I | exec: Running command: df --output=source,fstype
2016-12-09 08:41:13.091583 I | exec: Running command: lsblk /dev/sda6 -b -d -P -o SIZE,ROTA,RO,TYPE,PKNAME
2016-12-09 08:41:13.093435 I | exec: Running command: df --output=source,fstype
2016-12-09 08:41:13.100047 I | exec: Running command: lsblk /dev/sda7 -b -d -P -o SIZE,ROTA,RO,TYPE,PKNAME
2016-12-09 08:41:13.101817 I | exec: Running command: df --output=source,fstype
2016-12-09 08:41:13.114752 I | exec: Running command: lsblk /dev/sda9 -b -d -P -o SIZE,ROTA,RO,TYPE,PKNAME
2016-12-09 08:41:13.116594 I | exec: Running command: df --output=source,fstype
2016-12-09 08:41:13.125269 I | cephosd: Adding device sdb to desired state
2016-12-09 08:41:18.100895 I | cephmon: connecting to ceph cluster rookcluster with user admin
2016-12-09 08:41:18.157133 I | cephmon: connecting to ceph cluster rookcluster with user bootstrap-osd
2016-12-09 08:41:18.169577 I | clusterd: AGENT: Service osd succeeded the configuration
2016-12-09 08:41:19.162323 I | cephosd: successfully created OSD b15efabd-8318-40da-8b30-cf7e93a6899d with ID 1
2016-12-09 08:41:19.165236 I | exec: Running command: df --output=source,fstype
2016-12-09 08:41:19.167860 I | cephosd: Partitioning device sdb for bluestore
2016-12-09 08:41:19.173633 E | cephosd: failed to config osd on device sdb. failed device sdb. failed to partion device sdb. failed to get device sdb size. 100: Key not found (/rook/nodes/config/e57155ccfa35/disks/sdb) [238]
2016-12-09 08:41:19.173666 I | cephosd: 0/1 bluestore osds succeeded on this node
```

In addition to that, I also ran into #273 for a single pod (in a 4 node cluster)

**Comment by travisn** [comment]
thanks for opening this, upgrade hasn't been a priority yet for our pods while we are in an early preview mode. Submissions are welcome if you'd like to get involved!

**Comment by jbw976** [comment]
The design for upgrades in Rook as been merged with #983
